lib/ret/storage.ex

```diff
@@ -61,19 +72,62 @@ defmodule Ret.Storage do
   end
 
   def fetch(id, key) when is_binary(id) and is_binary(key) do
-    fetch_blob(id, key, @expiring_file_path)
+    fetch_blob(id, key, expiring_file_path())
   end
 
   def fetch(%OwnedFile{owned_file_uuid: id, key: key}) do
-    fetch_blob(id, key, @owned_file_path)
+    fetch_blob(id, key, owned_file_path())
+  end
+
+  def fetch(%CachedFile{} = cached_file) do
+    fetch_at(cached_file, Timex.now() |> Timex.to_naive_datetime() |> NaiveDateTime.truncate(:second))
+  end
+
+  defp maybe_bump_accessed_at(%CachedFile{accessed_at: accessed_at} = cached_file, time) do
+    one_day_ago = Timex.shift(time, days: -1)
+
+    if accessed_at < one_day_ago do
+      cached_file |> Ecto.Changeset.change(accessed_at: time) |> Ret.Repo.update()
+    end
+  end
+
+  def fetch_at(%CachedFile{file_uuid: id, file_key: key} = cached_file, time) do
+    maybe_bump_accessed_at(cached_file, time)
+
+    case fetch_blob(id, key, cached_file_path()) do
+      {:ok, meta, stream} ->
+        {:ok, meta, stream}
+
+      {:error, _} ->
+        migrate(%{cached_file: cached_file})
+        fetch_blob(id, key, cached_file_path())
+    end
+  end
+
+  # TODO: Remove this code once all the CachedFiles are stored in the cached_file_path().
+  defp migrate(%{
+         cached_file: %CachedFile{file_uuid: id, file_key: file_key} = cached_file
+       }) do
+    with {:ok, _meta, _stream} <- fetch_blob(id, file_key, expiring_file_path()),
+         {:ok, uuid} <- Ecto.UUID.cast(id),
+         [_src_file_directory, src_meta_file_path, src_blob_file_path] <- paths_for_uuid(uuid, expiring_file_path()),
+         [dest_file_directory, dest_meta_file_path, dest_blob_file_path] <- paths_for_uuid(uuid, cached_file_path()),
+         {:ok, _} <- File.mkdir_p(dest_file_directory),
+         {:ok, _} <- File.cp(src_meta_file_path, dest_meta_file_path),
+         {:ok, _} <- File.cp(src_blob_file_path, dest_blob_file_path) do
+      {:ok, cached_file}
+    else
+      _ -> {:error, "Migration failed"}
```

Comment:
Looks like this migration error is never handled anywhere. Should we at least log something here?

Reply:
I updated the code after your comment. Now this should return up through `Storage.fetch`, and then be handled by the caller. 
Do you think returning 500 errors [here](https://github.com/mozilla/reticulum/pull/481/files#diff-3a3972fecb3bde5b14ba6ca1228bb1f8500e1559159db614ece8faf314830a5dR204-R205) is a sufficiently good remedy, or should I also call [`Logger.error`](https://hexdocs.pm/logger/master/Logger.html#error/2)  ?

Reply:
I think we should almost never intentionally respond with a 500 status code. One of the other 400 statuses should be good, like [412](https://httpstatuses.com/412), or [424](https://httpstatuses.com/424)